test/server/config_validation/xds_verifier.cc

```diff
@@ -79,17 +89,25 @@ void XdsVerifier::listenerUpdated(const envoy::config::listener::v3::Listener& l
     return;
   }
 
-  for (unsigned long i = 0; i < listeners_.size(); ++i) {
-    const auto& rep = listeners_[i];
+  bool found = false;
+  for (auto it = listeners_.begin(); it != listeners_.end();) {
+    const auto& rep = *it;
+    ENVOY_LOG_MISC(debug, "checking {} for update", rep.listener.name());
     if (rep.listener.name() == listener.name()) {
-      if (rep.state == ACTIVE) {
+      // if we're updating a warming/active listener, num_modified_ must be incremented
+      if (rep.state != DRAINING && !found) {
         num_modified_++;
+        found = true;
+      }
+
+      if (rep.state == ACTIVE) {
         if (hasActiveRoute(listener)) {
           // if the new listener is ready to take traffic, the old listener will be removed
           // it seems to be directly removed without being added to the config dump as draining
           ENVOY_LOG_MISC(debug, "Removing {} after update", listener.name());
           num_active_--;
-          listeners_.erase(listeners_.begin() + i);
+          it = listeners_.erase(it);
```

Comment:
Some of these sub-cases are really interesting. I wonder if in a follow-up you may want to add a FAQ or elaborated documentation on some of these stats?
https://github.com/envoyproxy/envoy/blob/f64ae9c134794ca29b36e48158718a711d2fd4d3/docs/root/configuration/listeners/lds.rst

Reply:
I can definitely try and compile all the documentation I have across these files in one place and go through it with you or Harvey.  I'm not 100% confident I know the reasons behind why some of this behaviour happens, I found a lot of this out by just observing what Envoy is doing so I couldn't really talk too much on the rational behind some of it.  For example I'm not really sure why updating an active listener doesn't drain it, the docs seem to suggest that it should